README
  		README for L4/MIPS

Prequisites before building

1) You need mips-sgi-irix6 cross-compiler development tools.
	
2) A running version of GNU make, from here on referred to as make.



Now, to build the Downloadable Image Tool (DIT)

1) First set edit src/Makefile.conf as appropriate. In particular you
   will need to set the following according to your local environment:

    - TOP to the (absolute) path name of the directory of this README
      file;
    - TOOLBIN to the directory where tool binaries (which are to run on
      the development host) are to be installed;
    - INSTALL to the directory where L4/MIPS is to be installed (i.e.,
      the include files, libraries and kernel images);
    - TARGET_MACH to one of the supported machines;
    - XDEVBIN to where your cross development tools are installed;
    - HOST_BAUD to one of the baudrates;
    - BOOT to the directory used for booting;
    - KERNELNAME to the name under which the kernel image containing
      device drivers will be installed in $(INSTALL)/kernel.

   Some other variables may need defining if your enviroment has some
   unusual conventions (check MAKEWHATIS, MAKEWHATISFLAGS).


   The following are to be executed from $(TOP)/src:

2) Run "make depend" to build dependencies.

3) Run "make tools" to build dit.

4) Run "make tools-install" to install it in $(TOOLBIN)

Now to build L4/MIPS, serial driver and the libraries

1) run "make all"

2) run "make install"

You now have a bootable image (in $(INSTALL)/kernel/kernel-serial)
containing the kernel and serial driver, and a little example set up
ready for building in $(INSTALL)/example. See the docs for more info on
doing further development.

Kernel developers may find the programs in $(TOP)/src/kernel/test useful.

Good luck!

Distributed Systems Group,                 <disy@cse.unsw.edu.au>
School of Computer Science & Engineering   http://www.cse.unsw.edu.au/~disy
University of New South Wales,
Sydney 2052, Australia.


src/ChangeLog
2002-09-16  Cristan Szmajda <cls@cse.unsw.edu.au>

	* syscalls.S: committed k_id_nearest fix from `Inside L4' document
	
	* intr_serial.c: increased timeout to solve some serial problems
	
	* strcmp.S: made 64-bit clean

	* tagged whole thing as L4_81

2002-08-01  Cristan Szmajda <cls@cse.unsw.edu.au>

        * exc.S: added idt_fixup workaround for IDT 79R4600 rev. 2.0 erratum 7

2002-04-13  Daniel Potts <danielp@cse.unsw.edu.au> (cls)

	* updated DIT tool path
	
2000-08-22  Gernot Heiser <gernot@cse.unsw.EDU.AU>

	* Fixed u4600 serial driver code (matthewc@cse.unsw.edu.au).
	
2000-08-08  Gernot Heiser <gernot@zuse.disy.cse.unsw.EDU.AU>

	* Removed target "lib32" from master Makefile

	* tagged whole thing as L4_80
	
1999-10-28  Alan Au  <alanau@baragula.disy.cse.unsw.EDU.AU>

	* macros.h(wakeup_remaining): fixed incorrect operand order that
	was causing infinite kernel loop in l4_thread_schedule

1999-10-01  Alan Au  <alanau@baragula.disy.cse.unsw.EDU.AU>

	* debugger.c, panic.S: fix up register stacking for kdbg due to
	last fix

1999-09-30  Alan Au  <alanau@baragula.disy.cse.unsw.EDU.AU>

	* kernel.h, exc.S: fixed hi/lo bug (weren't stacked/unstacked before)

1999-09-29  Alan Au  <alanau@baragula.disy.cse.unsw.EDU.AU>

	* kernel.h, syscalls.S: patched in Daniel's fix for task delete

	* kernel/vm/vm-tlbcache-gpt-pair/gpt.c (gpt_print): enable
	printing of nodes 


1999-09-11  Gernot Heiser <gernot@zuse.disy.cse.unsw.EDU.AU>

	* Added kernel/test/reg_*.c testing register conservation

	* Added kernel/test/map_*.c testing mappings with large fpags

	* Fixed bug in kernel/vm/*/<PT>.c(vm_map): Send fpage of size 64
	lead to overflow resulting in no mapping at all.

	* Re-checked kernel panics and marked all as "ok" or "FIXME".
	
1999-09-03  Alan Au  <alanau@bassoon>

	* kernel/interrupts.S(int_ip7): Wakeup thread was not being
	inserted into the busy list. 

1999-07-20  Alan Au  <alanau@bassoon>

	* kernel/schedule.c(process_lists): fixed - removing tcb from busy
	list level that only contained one tcb broke the circular list

1999-04-14  Alan Au  <alanau@bassoon>

	* kernel/syscalls.S(k_thread_schedule): fixed a bug - was getting
	mcp from destination thread and not source task

1999-04-06  Alan Au  <alanau@baragula.disy.cse.unsw.edu.au>

	* Fixed schedule.c: wakeup thread should not preempt threads
	at the same priority.

Mon Mar 22 17:28:00 1999  Gernot Heiser <gernot@zuse.disy.cse.unsw.EDU.AU>
	
	* Fixed build on P4000i

	* Added <NL> -> <NL><CR> conversion in Indy serial driver
	
	* Added 32-bit ELF support to DIT (to support Linux)

	* Removed warnings in "make depend" from some Makefiles
	
Mon Mar  8 18:28:18 1999  Gernot Heiser <gernot@zuse.disy.cse.unsw.EDU.AU>

	* Added 32-bit library back again (to support Linux)

	* Luke's fix to make DIT cope with the elf flag indicating .noreorder
	
Wed Feb 10 15:52:52 1999  Gernot Heiser <gernot@zuse.disy.cse.unsw.EDU.AU>

	* tagged whole thing as L4_79
	
	* Added tests for chief IPC interfaces

	* Fixed bug in IPC path (kernel/exc.S) which overwrote original
	destination address in redirected IPC

	* Fixed bug in new l4_mips_ipc_chief_receive stub (lib/libl4/ipc.S)

	* Fixed Makefiles to stop on failed sub-make

	* Made build with either GNU or SGI tools (but SGI test fails)

	* Added "read" test, fixed "time" test.

	* Fixed time functions to account for 1ms accuracy of L4 timeouts.

1999-01-22  Alan Au  <alanau@baragula.disy.cse.unsw.edu.au>

	* Fixed bug in U4600 serial driver: cleared extra interrupts
	
Tue Dec 29 14:59:31 1998  Gernot Heiser <gernot@zuse.disy.cse.unsw.EDU.AU>
	
	* Fixed a race condition in kernel/drivers/intr_serial.c

	* Fixed a bug in kernel/drivers/serial.c.

	* Fixed return type of strlen() to make it compatible with gcc builtin

	* Added time.h, time.c in libl4, time.c in kernel tests.
	Time tester reads from serial port, should probably have a
	separate read tester.

	* Added missing style files in doc/l4uman/sty and updated
	Makefiles, dito unicrest-screen.eps.
	
	* Fixed Makefiles and example directory to be properly included in
	the installation.

	* Fixed src/include/Makefile to install kernel includes as well
	(at least dit.h is required)
	
Tue Dec 29 14:59:31 1998  Gernot Heiser <gernot@zuse.disy.cse.unsw.EDU.AU>
	
	* tagged whole thing as L4_78

	* Fixes to Makefiles & some sources to remove compiler warnings

Tue Dec 29 14:59:31 1998  Gernot Heiser <gernot@zuse.disy.cse.unsw.EDU.AU>
	
	* Incorporated docs into standard installation process

	* Made ref man grab kernel version from kernel sources

	* Renamed all *.ms to *.S

	* Eliminated all m4 use

	* Fixed declarations to eliminate almost all compiler warnings

Tue Dec 29 08:28:25 1998  Gernot Heiser <gernot@zuse.disy.cse.unsw.EDU.AU>
	
	* tagged whole thing as L4_77

	* Renamed src/lib64 to src/lib
	
Tue Dec 29 08:01:38 1998  Gernot Heiser <gernot@zuse.disy.cse.unsw.EDU.AU>

	* tagged whole thing as L4_77_tmp
	
	* Removed src/lib32, src/kernel/test32, src/tools/mk32

Mon Dec 28 23:20:41 1998  Gernot Heiser <gernot@zuse.disy.cse.unsw.EDU.AU>

	* tagged whole thing as L4_77_64_32

	* Renamed (old) lib to lib32

	* Fixed "make realclean", "make install", streamlined build
	process, incl making a driverless kernel as well as kernel-serial

	* Made kernel tests work again

	* Added sample outputs from kernel tests

	* Bug fix in lib64/ipc.ms

	* Removed a number of compiler warnings

	* Cleaned up some inconsistencies between header files in
	lib64/libc and include
	
Sat Dec 26 16:06:09 1998  Gernot Heiser <gernot@zuse.disy.cse.unsw.EDU.AU>

	* tagged whole thing as L4_77_64

	* Bug fixes in manuals

	* Support for "chief" type IPC interfaces, incl new man pages

	* Improved Makefiles ("make depend" works, separated installation
	from source directories)

	* Improved and clarified Installation procedure

	* Renamed some .ms to .S

	* Bug fixes in calling stubs (lib/lib64/ipc.ms)

	* Bug fixes in sigma0.c and kernel (to have consistent,

src/kernel/asid.S
/****************************************************************************
 *      $Id: asid.S,v 1.1 1998/12/29 05:22:09 gernot Exp $
 *      Copyright (C) 1997, 1998 Kevin Elphinstone, Univeristy of New South
 *      Wales.
 *
 *      This file is part of the L4/MIPS micro-kernel distribution.
 *
 *      This program is free software; you can redistribute it and/or
 *      modify it under the terms of the GNU General Public License
 *      as published by the Free Software Foundation; either version 2
 *      of the License, or (at your option) any later version.
 *      
 *      This program is distributed in the hope that it will be useful,
 *      but WITHOUT ANY WARRANTY; without even the implied warranty of
 *      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *      GNU General Public License for more details.
 *      
 *      You should have received a copy of the GNU General Public License
 *      along with this program; if not, write to the Free Software
 *      Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *      
 ****************************************************************************/

#include <asm.h>
#include <regdef.h>
#include <r4kc0.h>
#include <kernel/kernel.h>
#include <kernel/machine.h>
#include "macros.h"


/****************************************************************************
 *
 * Strategy for ASID management
 *
 * + ASIDs are stored in TCBs.
 *
 * + if ASID negative, then it's invalid.
 *
 * + A task's ASID is stored in thread 0, other threads get this one on demand.
 *
 * + ASIDs are kept in a free list.
 *
 * + if free list is empty ASIDs are removed from a task via FIFO arrangment.
 *
 ****************************************************************************/

PROC(asid_init)
	trace(asdi)
	move	a4, ra /* asid_free doesn't use a3 */
	lui	a0, KERNEL_BASE

	/* init free list */
	sd	zero, K_FREE_ASID_LIST(a0)

	/* init fifo */
	dli	a1, R4K_MAX_ASID
	sd	a1, K_ASID_FIFO_COUNT(a0)

	/* fill free list */
	dli	a3, 1 /* starting asid, 0 reserved by L4 */
	dli	a2, R4K_MAX_ASID + 1
	
1:	move	a0, a3
	jal	asid_free

	daddiu	a3, a3, 1

	bne	a2, a3, 1b
	
	jr	a4
END(asid_init)

PROC(asid_free)
	/* a0 has asid */
	.set	noat
	dli	t0, -1
	beq	t0, a0, 1f
	lui	t0, KERNEL_BASE
	dli	AT, ASID_BASE
	dsll	a0, a0, 3 
	daddu	a0, a0, AT
	ld	AT, K_FREE_ASID_LIST(t0)
	sd	a0, K_FREE_ASID_LIST(t0)
	sd	AT, (a0)
1:	jr	ra
	.set	at
END(asid_free)

PROC(asid_alloc)
	/* a0 has tid of thread requesting */
	.set	noat
	lui	t0, KERNEL_BASE
	ld	v0, K_FREE_ASID_LIST(t0)
	beq	v0, zero, 1f
	ld	AT, (v0)
	sd	a0, (v0)
	sd	AT, K_FREE_ASID_LIST(t0)
	dsrl	v0, v0, 3
	andi	v0, v0, 0xff
	jr	ra
1:	
	dli	v0, -1
	jr	ra	
	.set	at
END(asid_alloc)

.data
asid_get_msg:	
	.asciiz "no free ASIDS"	
PROC(asid_get)
	trace(asdg)
	.set	noat
	/* sp is valid and AT is only reg we can use */
	/* sp is for task needs ASID */

	/* first switch to asid 0 so tlb refills use system ASID */
	.set	noreorder
	dmtc0	zero, C0_ENTRYHI
	.set	reorder
	
	/* reserve some more registers to use */
	daddiu	sp, sp, -40
	sd	t0, (sp)
	sd	t1, 8(sp)
	sd	v0, 16(sp)
	sd	a0, 24(sp)
	sd	ra, 32(sp)

	
	/* check if task already has asid */
	tcbtop(t0)
	ld	t1, T_MYSELF-TCBO(t0)
	dsrl	a0, t1, TID_TASK_SHIFT
	dsll	a0, a0, TID_TASK_SHIFT
	tid2tcb(a0, t1)
	.set	noat
	ld	v0, T_ASID(t1)
	bltz	v0, 1f
	
	/* yes it does, so simply use that */
	.set	noreorder
	dmtc0	v0, C0_ENTRYHI
	b	2f
	sd	v0,  T_ASID-TCBO(t0)
	.set	reorder
	
1:
	/* else try to asid_alloc */

	jal	asid_alloc /* nails t0, v0, AT, ra */
	bltz	v0, 3f
	
	/* okay we has an asid from the free list */
	tcbtop(t0)
	.set noreorder
	sd	v0, T_ASID(t1)
	dmtc0	v0, C0_ENTRYHI
	b	2f
	sd	v0, T_ASID-TCBO(t0)
	.set	reorder	
3:

	/* Hmmm, no ASIDs available, need to remove one from active task */
	trace(asgt)
	/* pick asid from fifo counter */
	lui	t0, KERNEL_BASE
	ld	ra, K_ASID_FIFO_COUNT(t0)
	daddiu	a0, ra, -1
	bne	a0, zero, 4f
	dli	a0, R4K_MAX_ASID
4:	sd	a0, K_ASID_FIFO_COUNT(t0)
	trace(hvas)
	/* ra now has ASID, get tid that has it */
	dli	t0, ASID_BASE
	dsll	v0, ra, 3
	daddu	t0, t0, v0
	ld	t1, (t0)
	tcbtop(a0)
	ld	a0, T_MYSELF-TCBO(a0)
	sd	a0, (t0)   /* store my tid as current holder of asid */

	dsrl	t1, t1, TID_TASK_SHIFT
	dsll	t1, t1, TID_TASK_SHIFT /* AT now has taskid of old holder */
	tid2tcb(t1, a0) /* a0 now has tcb of old holder */
	.set	noat
	trace(nlas)
	/* store invalid ASID in task and threads of old holder */
	dli	AT, -1
1:	sd	AT, T_ASID(a0)
	ld	a0, T_PRESENT_NEXT(a0)
	bne	a0, zero, 1b
	
			
	/* TLB flush */
	trace(tlbf)
	move	a0, ra
	jal	tlb_flush_asid  /* uses AT, t0, t1, ra, preserves a0 */

#ifdef  TLB_CACHE
	trace(tb2f)
	/* TLB2 flush */
	jal	tlb_cache_flush_asid /* uses AT, t0, t1, ra, preserves a0 */
#endif	
	/* assign asid to current thread and thread 0 */
	trace(asas)
	tcbtop(t0)
	ld	t1, T_MYSELF-TCBO(t0)
	dsrl	v0, t1, TID_TASK_SHIFT
	dsll	v0, v0, TID_TASK_SHIFT
	tid2tcb(v0, t1)
	.set	noat
	sd	a0, T_ASID(t1)
	.set	noreorder
	dmtc0	a0, C0_ENTRYHI
	sd	a0, T_ASID-TCBO(t0)
	.set	reorder
	
2:
	ld	t0, (sp)
	ld	t1, 8(sp)
	ld	v0, 16(sp)
	ld	a0, 24(sp)
	ld	ra, 32(sp)
	daddiu	sp, sp, 40
	jr	ra
	.set	at
END(asid_get)
